File: 006_zigzag_conversion.cpp
/*
 Author: ZILI WANG
 Date: 15th Dec 2015
 */

class Solution {
public:
    string convert(string s, int numRows)
    {
        if (numRows == 1)
            return s;

        vector<vector<char>> matrix;

        for (unsigned int i  = 0; i < numRows; ++ i)
        {
            vector<char> row;
            matrix.push_back(row);
        }

        bool down = true;
        int idx = 0;
        bool finish = false;
        while (1)
        {
            if (finish) break;

            if (down)
            {
                for (unsigned int j = 0; j < numRows; ++ j)
                {
                    if (idx < s.size())
                    {
                        matrix[j].push_back(s[idx]);
                        ++ idx;
                    }
                    else
                    {
                        finish = true;
                        break;
                    }
                }
                down = false;
            }
            else
            {
                matrix[0].push_back(' ');
                matrix[numRows-1].push_back(' ');
                for (unsigned int j = numRows - 2; j > 0; -- j)
                {
                    if (idx < s.size())
                    {
                        matrix[j].push_back(s[idx]);
                        ++ idx;
                    }
                    else
                    {
                        finish = true;
                        break;
                    }
                }
                down = true;
            }
        }

        string output = "";
        for (size_t i = 0; i < matrix.size(); ++ i)
        {
            for (size_t j = 0; j < matrix[i].size(); ++ j)
            {
                if (matrix[i][j] != ' ')
                    output += matrix[i][j];
            }
        }

        return output;
    }
};

File: java/006_zigzag_conversion.java
/*
 * Author: Zili Wang
 * Date: 10th Feb 2016
 */

public class Solution {
    public String convert(String s, int numRows) {
        if (numRows == 1 || s.length() <= numRows) {
            return s;
        }
        
        String ans = "";
        ArrayList<ArrayList<Character>> m = new ArrayList<ArrayList<Character>>();
        for (int i = 0; i < numRows; ++ i) {
            ArrayList<Character> newList = new ArrayList<Character>();
            m.add(newList);
        }
        
        int i = 0;
        while (i < s.length()) {
            // from top to bottom
            for (int j = 0; j < numRows; ++j) {
                if (i >= s.length()) break;
                Character c = new Character(s.charAt(i));
                m.get(j).add(c);
                ++i;
            }
            
            // from bottom to top
            for (int j = numRows - 2; j > 0; --j) {
                if (i >= s.length()) break;
                Character c = new Character(s.charAt(i));
                m.get(j).add(c);
                ++i;
            }
        }
        
        for (i = 0; i < m.size(); ++i) {
            for (int j = 0; j < m.get(i).size(); ++j) {
                ans += m.get(i).get(j).toString();
            }
        }
        
        return ans;
    }
}
